### PR TITLE
[locator] pass/locate otInstance to all OpenThread classes

### PR DESCRIPTION
**Pull request by abtink**
This commit changes the model for OpenThread classes so that all
classes track/locate their owning/parent `otInstance` object (i.e.,
the owning `otInstance` object reference is expected to be passed  as
an input argument in the object constructor). This replaces and
simplifies current model where some of classes track/locate the parent
`ThreadNetif`, `MeshForwarder`, or `Ip6` object references instead.
This helps harmonize the model across OpenThread source and also
simplify the `Locator` class implementation.

### CODE REVIEW

File: src/core/common/locator.hpp [hunk 1]
```diff
@@ -44,197 +44,73 @@
...
-    }
-
-#if OPENTHREAD_ENABLE_MULTIPLE_INSTANCES
-    Type &mLocatorObject;
-#endif
-};
 
 /**
- * This class implements a locator for ThreadNetif object.
+ * This class implements locator for  otInstance object
```
Comment (jwhui) [inline src/core/common/locator.hpp]: nit: missing period.

**Comment by jwhui** [comment]
FWIW, this change seems to increase memory size marginally (about 256 bytes code size, 32 bytes RAM).

**Comment by abtink** [comment]
Interesting!.... I don't expect this to add any RAM...

**Comment by abtink** [comment]
There is a sutle and intresting reason behind the increase in RAM. Here is what I figured out:

Empty classes use 1 byte size to ensure they always get a unique address. "Empty base optimization" makes an exception for when an empty class is inherited and is used as base class. Our `InstanceLocator` is an empty base class (in the single instance case), so it's optimized and does not add any data.

On the other hand, C++ standard does not allow two objects of same type to share the same memory address. So if we have a class that inherits from `InstanceLocator` and then the first member variable defined in the class happens to be a type that also inherits from `InstanceLocator`, to ensure the two `InstanceLocator` objects do not share the same memory, some extra padding bytes will be added by the complier (please see [empty base optimization](http://en.cppreference.com/w/cpp/language/ebo)).

`Ip6:Mpl` is an example for this case: It is itself an `InstanceLocator` and the first data member variable is a `TimerMilli mSeedSetTimer` which again inherits from `InstanceLocator`.

We can avoid this by re-arranging the member variable declarations (when possible).
Alternatively, we can change `TimerMilli/Usec` or `Tasklet` (which are the more commonly used member variable types) to inherit from a different locator class type

File: src/core/common/locator.hpp (hunk 1 above)
Comment (abtink) [reply src/core/common/locator.hpp]: Fixed it in new pushed commit.

**Comment by abtink** [comment]
Removing `work-in-progress`. About the small RAM increase, I think it'd ok for now. We can address it later by re-arranging the member variable declaration (in a separate PR). Would be good to use "group by alignment" or "group by size" as a recommended policy for member variables..

**Comment by abtink** [comment]
This PR now contains two commits one containing the original change, second one reorders the member variable declarations. @jwhui is it possible to merge this PR without squashing the two commit? If not, I suggest to separate them into two PRs...Thoughts?

**Pull request merged**